src/mococrw/symmetric_crypto.h

```diff
@@ -0,0 +1,366 @@
+    virtual ~SymmetricCipherI() = default;
+
+    /**
+     * Decrypt or encrypt chunk of data.
+     *
+     * This function encrypts or decrypts chunk of data and places result in the internal buffer.
+     * Processed data can be accessed via read() or/and readAll().
+     *
+     * @param message chunk of data to encrypt/decrypt.
+     */
+    virtual void update(const std::vector<uint8_t> &message) = 0;
+
+    /**
+     * Read portion of encrypted/decrypted data from cipher buffer.
+     *
+     *
+     * Depending on memory strategy and size of read chunk, performance of this method vary. The
+     * current version uses queue of vectors as internal buffer. Therefore, alternate update() and
+     * read() of data blocks of the same size are zero-copy operations.
+     *
+     * @param length size of the requested chunk of data.
+     * @return processed chunk of \c length or smaller.
+     */
+    virtual std::vector<uint8_t> read(size_t length) = 0;
+
+    /**
+     * Read all available data encrypted/decrypted data from the cipher buffer.
+     *
+     * In the current implementation, readAll() reassembles the result chunk of data from pieces,
+     * previously provided to update(). It requires additional copy operationscomparing to update(),
+     * read(), update(), read().
+     *
+     * @return all processed data
+     */
+    virtual std::vector<uint8_t> readAll() = 0;
+
+    /**
+     * Finalize encryption/decryption and return all processed data not yet read by read()/readAll().
+     *
+     * @return
+     */
+    virtual std::vector<uint8_t> finish() = 0;
+
+    /**
+     * Get initialization vector
+     *
+     * @sa AESCipherBuilder
+     *
+     * @return
+     */
+    virtual std::vector<uint8_t> getIV() const = 0;
+};
+
+/**
+ * This interface declares an authenticated encryption extension to symmetric cipher.
+ *
+ * @code
+ *   // Encryption
+ *   auto encryptor = AESCipherBuilder{SymmetricCipherMode::GCM, SymmetricCipherKeySize::S_256, secretKey}.buildAuthenticatedEncryptor();
+ *   encryptor->update(plaintext);
+ *   auto ciphertext = encryptor->finish();
+ *   auto iv = encryptor->getIV();
+ *   auto tag = encryptor->getAuthTag();
+ *
+ *   // Decryption
+ *   auto decryptor = AESCipherBuilder{SymmetricCipherMode::GCM, secretKey}.setIV(iv).buildAuthenticatedDecryptor();
+ *   decryptor->update(ciphertext);
+ *   decryptor->setAuthTag(tag);
+ *   auto decryptedText = decryptor->finish();
+ * @endcode
+ */
+class AuthenticatedEncryptionI
+{
+public:
+    /**
+     * Set authentication tag.
+     *
+     * For an authenticated cipher, setAuthTag() must be called before finalyzing decryption
+     * SymmetricCipherI::finish().
+     *
+     * @param tag authentication tag
+     */
+    virtual void setAuthTag(const std::vector<uint8_t> &tag) = 0;
+
+    /**
+     * Get authentication tag.
+     *
+     * For an authenticated cipher, getAuthTag() must be after encryption is finalized (SymmetricCipherI::finish()).
```

Comment:
I think this sentence needs "called" as another verb.

Also, we should really document how failures in decryption/encryption and/or verification of the auth tag manifest themselves. IIRC we do throw exceptions for those, and that should probably be documented here.

Reply:
"called" - fixed.
Exceptions - agree, but problem is OpenSSL does not return you any specific error details for authenticated encryption. Can we do it in a separate patch as this MR gets unmanageable?